File: native/groupcontrol/carriers.go
// Package groupcontrol encodes private proof carriers. Outer signatures do not
// confer membership authority; creator certificates are verified by the registry.
package groupcontrol

import (
	"bytes"
	"encoding/json"
	"errors"
	"sort"
	"strconv"
	"time"

	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/groupauthority"
	"github.com/JohnnyPBelo/relayloom/native/groups"
)

const PlainLimit = 1024 * 1024
const BundleLimit = 2 * 1024 * 1024
const TTL = int64(3600000)

type InputError struct{ Message string }

func (e *InputError) Error() string { return e.Message }
func bad(message string) error      { return &InputError{message} }

type Authorization struct {
	Number     int
	ID         string
	Invitation any
}
type Control struct {
	Raw                          map[string]any
	Action, GroupID, To, EpochID string
	From, Count, Number          int
	Authorization                Authorization
	Headers                      []any
	Head, Epoch, Snapshot        any
}

func str(v any) string { s, _ := v.(string); return s }
func integer(v any) (int, bool) {
	var n int64
	switch v := v.(type) {
	case int:
		n = int64(v)
	case int64:
		n = v
	case float64:
		n = int64(v)
		if float64(n) != v {
			return 0, false
		}
	case json.Number:
		var e error
		n, e = strconv.ParseInt(string(v), 10, 64)
		if e != nil {
			return 0, false
		}
	default:
		return 0, false
	}
	return int(n), n >= 0 && n < 1024
}
func exact(v map[string]any, fields ...string) bool {
	if len(v) != len(fields) {
		return false
	}
	for _, key := range fields {
		if _, ok := v[key]; !ok {
			return false
		}
	}
	return true
}
func bounded(value any, limit int) error {
	data, err := core.Canonical(value)
	if err != nil {
		return bad(err.Error())
	}
	if len(data) > limit {
		return bad("carrier excede limite")
	}
	return nil
}
func equal(a, b any) bool {
	x, e := core.Canonical(a)
	y, f := core.Canonical(b)
	return e == nil && f == nil && bytes.Equal(x, y)
}
func ids(cards []core.PublicIdentity) []string {
	out := []string{}
	for _, card := range cards {
		out = append(out, card.ID)
	}
	sort.Strings(out)
	return out
}
func readers(bundle core.Bundle) []string {
	out := []string{}
	for _, key := range bundle.Manifest.Keys {
		out = append(out, key.Reader)
	}
	sort.Strings(out)
	return out
}

// Parse checks structure/resources, not every certificate in a page. Valid
// restrictive prefixes must reach the registry even if a later proof is bad.
func Parse(value any) (Control, error) {
	var c Control
	if err := bounded(value, PlainLimit); err != nil {
		return c, err
	}
	data, err := core.Canonical(value)
	if err != nil {
		return c, err
	}
	decoded, err := core.DecodeJSON(data, PlainLimit)
	if err != nil {
		return c, bad(err.Error())
	}
	v, ok := decoded.(map[string]any)
	if !ok {
		return c, bad("carrier inválido")
	}
	version, ok := integer(v["version"])
	if !ok || version != 1 || str(v["type"]) != "group-control" || !core.ValidAddress(str(v["groupId"])) {
		return c, bad("carrier inválido")
	}
	c.Raw, c.Action, c.GroupID = v, str(v["action"]), str(v["groupId"])
	common := []string{"type", "version", "groupId", "action"}
	switch c.Action {
	case "headers-request":
		if !exact(v, append(common, "to", "from", "count", "authorization")...) {
			return c, bad("campos de pedido inválidos")
		}
		c.To = str(v["to"])
		c.From, ok = integer(v["from"])
		count, countOK := integer(v["count"])
		c.Count = count
		auth, authOK := v["authorization"].(map[string]any)
		if !core.ValidAddress(c.To) || !ok || !countOK || count < 1 || count > groupauthority.PageLimit || !authOK {
			return c, bad("pedido de headers inválido")
		}
		fields := []string{"number", "id"}
		if _, has := auth["invitation"]; has {
			fields = append(fields, "invitation")
		}
		if !exact(auth, fields...) {
			return c, bad("autorização inválida")
		}
		c.Authorization.Number, ok = integer(auth["number"])
		c.Authorization.ID = str(auth["id"])
		c.Authorization.Invitation = auth["invitation"]
		if !ok || !core.ValidAddress(c.Authorization.ID) {
			return c, bad("autorização inválida")
		}
		if _, has := auth["invitation"]; has {
			if err = bounded(auth["invitation"], groups.CertificateBytes); err != nil {
				return c, err
			}
		}
	case "snapshot-request":
		if !exact(v, append(common, "to", "number", "epochId")...) {
			return c, bad("campos de pedido inválidos")
		}
		c.To, c.EpochID = str(v["to"]), str(v["epochId"])
		c.Number, ok = integer(v["number"])
		if !ok || !core.ValidAddress(c.To) || !core.ValidAddress(c.EpochID) {
			return c, bad("pedido de snapshot inválido")
		}
	case "headers":
		if !exact(v, append(common, "to", "from", "headers", "head")...) {
			return c, bad("campos de headers inválidos")
		}
		c.To = str(v["to"])
		c.From, ok = integer(v["from"])
		var headersOK bool
		c.Headers, headersOK = v["headers"].([]any)
		c.Head = v["head"]
		if !core.ValidAddress(c.To) || !ok || !headersOK || len(c.Headers) > groupauthority.PageLimit {
			return c, bad("página inválida")
		}
		if err = bounded(c.Headers, groupauthority.PageBytes); err != nil {
			return c, err
		}
		if err = bounded(c.Head, groups.HeaderBytes); err != nil {
			return c, err
		}
		for _, header := range c.Headers {
			if err = bounded(header, groups.HeaderBytes); err != nil {
				return c, err
			}
		}
	case "snapshot":
		fields := append(common, "epoch", "snapshot")
		if _, present := v["to"]; present {
			fields = append(fields, "to")
			c.To = str(v["to"])
			if !core.ValidAddress(c.To) {
				return c, bad("destinatário inválido")
			}
		}
		if !exact(v, fields...) {
			return c, bad("campos de snapshot inválidos")
		}
		c.Epoch, c.Snapshot = v["epoch"], v["snapshot"]
		if err = bounded(c.Epoch, groups.HeaderBytes); err != nil {
			return c, err
		}
		if err = bounded(c.Snapshot, groups.SnapshotBytes); err != nil {
			return c, err
		}
	default:
		return c, bad("acção de controlo desconhecida")
	}
	return c, nil
}

// ManifestPolicy is for already verified manifests. Ciphertext has the same
// length as plaintext (the GCM tag is separate), even for an opaque seeder.
func ManifestPolicy(m core.Manifest) bool {
	bytes := 0
	for _, c := range m.Chunks {
		bytes += c.Size
	}
	return m.Kind == "group-control" && m.PublicKey == nil && m.Expires-m.Created <= TTL+10 && bytes <= PlainLimit
}

func VerifyEnvelope(bundle core.Bundle) error {
	if err := bounded(bundle, BundleLimit); err != nil {
		return err
	}
	if err := core.VerifyBundle(bundle); err != nil {
		return bad(err.Error())
	}
	if !ManifestPolicy(bundle.Manifest) || bundle.Manifest.Expires <= time.Now().UnixMilli() {
		return bad("controlo exige envelope privado limitado e não expirado")
	}
	return nil
}

func Open(bundle core.Bundle, identity core.Identity) (Control, error) {
	var c Control
	if err := VerifyEnvelope(bundle); err != nil {
		return c, err
	}
	value, err := core.DecryptBundle(bundle, &identity)
	if err != nil {
		return c, bad(err.Error())
	}
	c, err = Parse(value)
	if err != nil {
		return c, err
	}
	if c.To != "" {
		if c.To != identity.Public.ID && bundle.Manifest.Author.ID != identity.Public.ID {
			return c, bad("carrier dirigido a outra identidade")
		}
		expected := []string{c.To}
		if c.To != bundle.Manifest.Author.ID {
			expected = append(expected, bundle.Manifest.Author.ID)
		}
		sort.Strings(expected)
		if !equal(readers(bundle), expected) {
			return c, bad("envelope de controlo alarga audiência")
		}
	}
	return c, nil
}
func Seal(identity core.Identity, payload map[string]any, recipients []core.PublicIdentity) (core.Bundle, error) {
	var bundle core.Bundle
	c, err := Parse(payload)
	if err != nil {
		return bundle, err
	}
	bundle, err = core.CreateBundle(identity, "group-control", c.Raw, recipients, false, TTL)
	if err != nil {
		return bundle, err
	}
	verified, err := Open(bundle, identity)
	if err != nil {
		return bundle, err
	}
	if !equal(c.Raw, verified.Raw) {
		return bundle, errors.New("controlo assinado incoerente")
	}
	return bundle, nil
}
func HasCard(epoch groups.GroupEpoch, card core.PublicIdentity) bool {
	hash, err := groups.MemberCardHash(card)
	if err != nil {
		return false
	}
	for _, member := range epoch.Body.Members {
		if member.ID == card.ID && member.CardHash == hash {
			return true
		}
	}
	return false
}
func retained(g *groupauthority.Registry, id string, number int, epochID string, allowFork bool) (groups.GroupEpoch, error) {
	var result groups.GroupEpoch
	if _, err := g.ScopeGeneration(); err != nil {
		return result, err
	}
	headers, err := g.Proofs(id, number, 1)
	if err != nil {
		return result, err
	}
	if len(headers) > 0 && headers[0].ID == epochID {
		return headers[0], nil
	}
	if allowFork {
		stop, err := g.StopEvidence(id)
		if err != nil {
			return result, err
		}
		if stop != nil && stop.Reason == "forked" {
			for _, epoch := range []*groups.GroupEpoch{stop.First, stop.Second} {
				if epoch != nil && epoch.Body.Number == number && epoch.ID == epochID {
					return *epoch, nil
				}
			}
		}
	}
	return result, bad("época de autorização indisponível")
}
func privateEpoch(g *groupauthority.Registry, epoch groups.GroupEpoch) error {
	stop, err := g.StopEvidence(epoch.Body.GroupID)
	if err != nil {
		return err
	}
	if stop != nil && stop.Reason == "forked" && epoch.Body.Number >= stop.First.Body.Number {
		return bad("snapshot de época em conflito não pode ser redistribuído")
	}
	return nil
}

type HeaderResult struct {
	Headers []groups.GroupEpoch
	Head    groups.GroupEpoch
}

func RequestedHeaders(g *groupauthority.Registry, c Control, requester core.PublicIdentity) (HeaderResult, error) {
	var result HeaderResult
	parsed, parseErr := Parse(c.Raw)
	if parseErr != nil {
		return result, parseErr
	}
	c = parsed
	if c.Action != "headers-request" {
		return result, bad("pedido inválido")
	}
	anchor, err := g.Anchor(c.GroupID)
	if err != nil {
		return result, err
	}
	auth, err := retained(g, c.GroupID, c.Authorization.Number, c.Authorization.ID, true)
	if err != nil {
		return result, err
	}
	end := auth
	if _, present := c.Raw["authorization"].(map[string]any)["invitation"]; present {
		raw, err := core.Canonical(c.Authorization.Invitation)
		if err != nil {
			return result, bad(err.Error())
		}
		if _, err = groups.DecodeInvitation(raw, anchor, auth, requester); err != nil {
			return result, bad(err.Error())
		}
		next, err := g.Proofs(c.GroupID, min(1023, auth.Body.Number+1), 1)
		if err != nil {
			return result, err
		}
		if len(next) > 0 {
			end = next[0]
		}
	} else {
		if !HasCard(auth, requester) {
			return result, bad("solicitante não pertence à época")
		}
		cursor, stopped := auth.Body.Number+1, auth.Body.State == "closed"
		for !stopped && cursor < 1024 {
			page, err := g.Proofs(c.GroupID, cursor, groupauthority.PageLimit)
			if err != nil {
				return result, err
			}
			if len(page) == 0 {
				break
			}
			for _, header := range page {
				end = header
				cursor = header.Body.Number + 1
				if header.Body.State == "closed" || !HasCard(header, requester) {
					stopped = true
					break
				}
			}
		}
	}
	result.Head = end
	result.Headers = []groups.GroupEpoch{}
	if c.From <= end.Body.Number {
		result.Headers, err = g.Proofs(c.GroupID, c.From, min(c.Count, end.Body.Number-c.From+1))
		if err != nil {
			return result, err
		}
	}
	conflict, err := g.StopEvidence(c.GroupID)
	if err != nil {
		return result, err
	}
	if conflict != nil && conflict.Reason == "forked" && conflict.First.Body.Number <= end.Body.Number {
		extra := []groups.GroupEpoch{*conflict.First, *conflict.Second}
		if extra[0].ID == auth.ID {
			extra[0], extra[1] = extra[1], extra[0]
		}
		for _, header := range extra {
			found := false
			for _, item := range result.Headers {
				found = found || item.ID == header.ID
			}
			if found {
				continue
			}
			candidate := append(append([]groups.GroupEpoch{}, result.Headers...), header)
			if len(candidate) > c.Count || bounded(candidate, groupauthority.PageBytes) != nil {
				break
			}
			result.Headers = candidate
		}
	}
	return result, nil
}

type SnapshotResult struct {
	Epoch    groups.GroupEpoch
	Snapshot groups.GroupSnapshot
}

func RequestedSnapshot(g *groupauthority.Registry, c Control, requester, producer core.PublicIdentity) (SnapshotResult, error) {
	var result SnapshotResult
	parsed, parseErr := Parse(c.Raw)
	if parseErr != nil {
		return result, parseErr
	}
	c = parsed
	if c.Action != "snapshot-request" {
		return result, bad("pedido inválido")
	}
	epoch, err := retained(g, c.GroupID, c.Number, c.EpochID, false)
	if err != nil {
		return result, err
	}
	if !HasCard(epoch, requester) || !HasCard(epoch, producer) {
		return result, bad("snapshot não autoriza estes cartões")
	}
	if err = privateEpoch(g, epoch); err != nil {
		return result, err
	}
	snapshot, err := g.PrivateState(c.GroupID, epoch.ID)
	if err != nil {
		return result, err
	}
	if snapshot == nil {
		return result, bad("snapshot privado indisponível")
	}
	result = SnapshotResult{epoch, *snapshot}
	return result, nil
}

type Material struct {
	Payload map[string]any
	Readers []core.PublicIdentity
}

func SnapshotMaterial(g *groupauthority.Registry, producer core.PublicIdentity, epoch groups.GroupEpoch, snapshot groups.GroupSnapshot, recipient ...core.PublicIdentity) (Material, error) {
	var result Material
	if _, err := g.ScopeGeneration(); err != nil {
		return result, err
	}
	anchor, err := g.Anchor(epoch.Body.GroupID)
	if err != nil {
		return result, err
	}
	if err = groups.VerifyEpoch(epoch, anchor); err != nil {
		return result, err
	}
	if err = groups.VerifySnapshot(snapshot, anchor, epoch); err != nil {
		return result, err
	}
	if err = privateEpoch(g, epoch); err != nil {
		return result, err
	}
	if _, err = retained(g, anchor.ID, epoch.Body.Number, epoch.ID, false); err != nil {
		return result, err
	}
	if !HasCard(epoch, producer) {
		return result, bad("só leitor autorizado pode transportar snapshot")
	}
	result = Material{map[string]any{"type": "group-control", "version": 1, "action": "snapshot", "groupId": anchor.ID, "epoch": epoch, "snapshot": snapshot}, snapshot.Members}
	if len(recipient) > 1 {
		return result, bad("destinatário inválido")
	}
	if len(recipient) == 1 {
		if !HasCard(epoch, recipient[0]) {
			return result, bad("destinatário fora do snapshot")
		}
		result.Payload["to"] = recipient[0].ID
		result.Readers = []core.PublicIdentity{recipient[0]}
	}
	return result, nil
}
func SnapshotPreview(bundle core.Bundle, c Control, anchor groups.GroupAnchor) (SnapshotResult, error) {
	var result SnapshotResult
	if c.Action != "snapshot" {
		return result, bad("snapshot inválido")
	}
	raw, err := core.Canonical(c.Epoch)
	if err != nil {
		return result, bad(err.Error())
	}
	epoch, err := groups.DecodeEpoch(raw, anchor)
	if err != nil {
		return result, bad(err.Error())
	}
	raw, err = core.Canonical(c.Snapshot)
	if err != nil {
		return result, bad(err.Error())
	}
	snapshot, err := groups.DecodeSnapshot(raw, anchor, epoch)
	if err != nil {
		return result, bad(err.Error())
	}
	expected := ids(snapshot.Members)
	if c.To != "" {
		found := false
		for _, m := range epoch.Body.Members {
			found = found || m.ID == c.To
		}
		if !found {
			return result, bad("destinatário fora do snapshot")
		}
		expected = []string{c.To}
		if c.To != bundle.Manifest.Author.ID {
			expected = append(expected, bundle.Manifest.Author.ID)
		}
		sort.Strings(expected)
	}
	if !HasCard(epoch, bundle.Manifest.Author) || !equal(readers(bundle), expected) {
		return result, bad("carrier de snapshot alarga leitores ou assinante")
	}
	result = SnapshotResult{epoch, snapshot}
	return result, nil
}
func CheckedSnapshot(g *groupauthority.Registry, bundle core.Bundle, c Control) (SnapshotResult, error) {
	var result SnapshotResult
	if _, err := g.ScopeGeneration(); err != nil {
		return result, err
	}
	anchor, err := g.Anchor(c.GroupID)
	if err != nil {
		return result, err
	}
	result, err = SnapshotPreview(bundle, c, anchor)
	if err != nil {
		return result, err
	}
	if _, err = retained(g, c.GroupID, result.Epoch.Body.Number, result.Epoch.ID, false); err != nil {
		return result, err
	}
	if err = privateEpoch(g, result.Epoch); err != nil {
		return result, err
	}
	return result, nil
}

File: packages/groups/src/carriers.ts
import {
  canonical,
  createBundle,
  decryptBundle,
  verifyBundle,
  type Bundle,
  type Identity,
  type PublicIdentity,
  type Manifest,
} from "../../core/src/index.js";
import {
  memberCardHash,
  verifyGroupEpoch,
  verifyGroupInvitation,
  verifyGroupSnapshot,
  type GroupEpoch,
  type GroupInvitation,
  type GroupSnapshot,
} from "./certificates.js";
import type { GroupRegistry } from "./registry.js";

export const CONTROL_LIMITS = Object.freeze({
  plaintext: 1024 * 1024,
  bundle: 2 * 1024 * 1024,
  page: 16,
  pageBytes: 256 * 1024,
  ttl: 3600_000,
});
export class GroupControlError extends Error {}
export type ControlAuthorization = {
  number: number;
  id: string;
  invitation?: GroupInvitation;
};
type Common = { type: "group-control"; version: 1; groupId: string };
export type GroupControl =
  | (Common & {
      action: "headers-request";
      to: string;
      from: number;
      count: number;
      authorization: ControlAuthorization;
    })
  | (Common & {
      action: "snapshot-request";
      to: string;
      number: number;
      epochId: string;
    })
  | (Common & {
      action: "headers";
      to: string;
      from: number;
      headers: unknown[];
      head: unknown;
    })
  | (Common & {
      action: "snapshot";
      epoch: unknown;
      snapshot: unknown;
      to?: string;
    });

const address = (v: unknown): v is string =>
  typeof v === "string" && /^[a-f0-9]{64}$/.test(v);
const number = (v: unknown): v is number =>
  Number.isInteger(v) && Number(v) >= 0 && Number(v) < 1024;
function insist(
  value: unknown,
  message = "Carrier de grupo inválido",
): asserts value {
  if (!value) throw new GroupControlError(message);
}
function input<T>(fn: () => T): T {
  try {
    return fn();
  } catch (error) {
    throw new GroupControlError((error as Error).message);
  }
}
function exact(value: any, keys: string[]) {
  insist(
    value &&
      typeof value === "object" &&
      !Array.isArray(value) &&
      Object.keys(value).sort().join(",") === [...keys].sort().join(","),
  );
}
function size(value: unknown, limit: number) {
  insist(
    Buffer.byteLength(canonical(value)) <= limit,
    "Carrier excede o limite",
  );
}

/** Shape/resource validation only. In particular, do not reject a whole page
 * for a bad trailing certificate before the registry commits its valid fence. */
export function parseGroupControl(value: unknown): GroupControl {
  size(value, CONTROL_LIMITS.plaintext);
  const v = value as any;
  insist(v?.type === "group-control" && v.version === 1 && address(v.groupId));
  const common = ["type", "version", "groupId", "action"];
  switch (v.action) {
    case "headers-request":
      exact(v, [...common, "to", "from", "count", "authorization"]);
      insist(
        address(v.to) &&
          number(v.from) &&
          Number.isInteger(v.count) &&
          v.count >= 1 &&
          v.count <= CONTROL_LIMITS.page,
      );
      exact(v.authorization, [
        "number",
        "id",
        ...(Object.hasOwn(v.authorization ?? {}, "invitation")
          ? ["invitation"]
          : []),
      ]);
      insist(number(v.authorization.number) && address(v.authorization.id));
      if (v.authorization.invitation !== undefined)
        size(v.authorization.invitation, 2048);
      break;
    case "snapshot-request":
      exact(v, [...common, "to", "number", "epochId"]);
      insist(address(v.to) && number(v.number) && address(v.epochId));
      break;
    case "headers":
      exact(v, [...common, "to", "from", "headers", "head"]);
      insist(
        address(v.to) &&
          number(v.from) &&
          Array.isArray(v.headers) &&
          v.headers.length <= CONTROL_LIMITS.page,
      );
      size(v.headers, CONTROL_LIMITS.pageBytes);
      size(v.head, 16384);
      for (const header of v.headers) size(header, 16384);
      break;
    case "snapshot":
      exact(v, [
        ...common,
        "epoch",
        "snapshot",
        ...(Object.hasOwn(v, "to") ? ["to"] : []),
      ]);
      if (Object.hasOwn(v, "to")) insist(address(v.to));
      size(v.epoch, 16384);
      size(v.snapshot, 524288);
      break;
    default:
      throw new Error("Acção de controlo desconhecida");
  }
  return structuredClone(v);
}

export function openGroupControl(
  bundle: Bundle,
  identity: Identity,
): GroupControl {
  verifyGroupControlEnvelope(bundle);
  const payload = parseGroupControl(decryptBundle(bundle, identity));
  const readers = bundle.manifest.keys.map((k) => k.reader).sort();
  if ("to" in payload) {
    insist(
      payload.to === identity.public.id ||
        bundle.manifest.author.id === identity.public.id,
      "Carrier dirigido a outra identidade",
    );
    insist(
      canonical(readers) ===
        canonical([...new Set([payload.to, bundle.manifest.author.id])].sort()),
      "Envelope de controlo alarga a audiência",
    );
  }
  return payload;
}

/** Policy over an already signature-verified manifest, usable by an opaque
 * seeder. AES-GCM stores its tag separately, so chunk bytes bound plaintext. */
export function groupControlPolicy(manifest: Manifest): boolean {
  return (
    manifest.kind === "group-control" &&
    !manifest.publicKey &&
    manifest.expires - manifest.created <= CONTROL_LIMITS.ttl + 10 &&
    manifest.chunks.reduce((n, c) => n + c.size, 0) <= CONTROL_LIMITS.plaintext
  );
}

/** Reject at the routing validation boundary, before opaque forwarding/store. */
export function verifyGroupControlEnvelope(bundle: Bundle): void {
  size(bundle, CONTROL_LIMITS.bundle);
  verifyBundle(bundle);
  insist(
    bundle.manifest.kind === "group-control" && !bundle.manifest.publicKey,
    "Controlo exige um envelope privado",
  );
  insist(bundle.manifest.expires > Date.now(), "Carrier expirado");
  insist(
    groupControlPolicy(bundle.manifest),
    "Envelope de controlo excede duração ou tamanho",
  );
}

/** Low-level sealing is not membership authority. Runtime callers must first
 * derive recipients/proofs from the authenticated registry transaction. */
export function sealGroupControl(
  identity: Identity,
  value: GroupControl,
  recipients: PublicIdentity[],
): Bundle {
  const payload = parseGroupControl(value);
  const bundle = createBundle(
    identity,
    "group-control",
    payload,
    recipients,
    CONTROL_LIMITS.ttl,
  );
  size(bundle, CONTROL_LIMITS.bundle);
  verifyBundle(bundle);
  insist(
    canonical(openGroupControl(bundle, identity)) === canonical(payload),
    "Controlo assinado incoerente",
  );
  return bundle;
}

function hasCard(epoch: GroupEpoch, card: PublicIdentity) {
  return epoch.body.members.some(
    (m) => m.id === card.id && m.cardHash === memberCardHash(card),
  );
}
function retainedEpoch(
  g: GroupRegistry,
  groupId: string,
  n: number,
  id: string,
  allowFork = false,
) {
  g.assertScope();
  let epoch = g.proofs(groupId, n, 1)[0];
  if (allowFork && epoch?.id !== id) {
    const evidence = g.stopEvidence(groupId);
    if (evidence?.reason === "forked")
      epoch = [evidence.first, evidence.second].find(
        (e) => e.body.number === n && e.id === id,
      )!;
  }
  insist(epoch && epoch.id === id, "Época de autorização indisponível");
  return epoch;
}

/** A removed member can learn the boundary that removed its committed card,
 * but not use an ancient membership to enumerate all later memberships. */
export function requestedHeaders(
  g: GroupRegistry,
  request: Extract<GroupControl, { action: "headers-request" }>,
  requester: PublicIdentity,
) {
  parseGroupControl(request);
  const anchor = g.anchor(request.groupId);
  const auth = retainedEpoch(
    g,
    request.groupId,
    request.authorization.number,
    request.authorization.id,
    true,
  );
  let end = auth;
  if (Object.hasOwn(request.authorization, "invitation")) {
    input(() =>
      verifyGroupInvitation(
        request.authorization.invitation,
        anchor,
        auth,
        requester,
      ),
    );
    end =
      g.proofs(request.groupId, Math.min(1023, auth.body.number + 1), 1)[0] ??
      auth;
  } else {
    insist(hasCard(auth, requester), "Solicitante não pertence à época");
    let cursor = auth.body.number + 1,
      stopped = auth.body.state === "closed";
    while (!stopped && cursor < 1024) {
      const page = g.proofs(request.groupId, cursor, CONTROL_LIMITS.page);
      if (!page.length) break;
      for (const header of page) {
        end = header;
        cursor = header.body.number + 1;
        if (header.body.state === "closed" || !hasCard(header, requester)) {
          stopped = true;
          break;
        }
      }
    }
  }
  const headers =
    request.from > end.body.number
      ? []
      : g.proofs(
          request.groupId,
          request.from,
          Math.min(request.count, end.body.number - request.from + 1),
        );
  const conflict = g.stopEvidence(request.groupId);
  if (
    conflict?.reason === "forked" &&
    conflict.first.body.number <= end.body.number
  ) {
    const alternatives = [conflict.first, conflict.second].sort(
      (a, b) => Number(a.id === auth.id) - Number(b.id === auth.id),
    );
    for (const header of alternatives) {
      if (headers.some((h) => h.id === header.id)) continue;
      if (
        headers.length >= request.count ||
        Buffer.byteLength(canonical([...headers, header])) >
          CONTROL_LIMITS.pageBytes
      )
        break;
      headers.push(header);
    }
  }
  return { headers, head: end };
}

function privateEpoch(g: GroupRegistry, epoch: GroupEpoch) {
  const evidence = g.stopEvidence(epoch.body.groupId);
  insist(
    evidence?.reason !== "forked" ||
      epoch.body.number < evidence.first.body.number,
    "Snapshot de uma época em conflito não pode ser redistribuído",
  );
}

export function requestedSnapshot(
  g: GroupRegistry,
  request: Extract<GroupControl, { action: "snapshot-request" }>,
  requester: PublicIdentity,
  producer: PublicIdentity,
) {
  parseGroupControl(request);
  const epoch = retainedEpoch(
    g,
    request.groupId,
    request.number,
    request.epochId,
  );
  insist(
    hasCard(epoch, requester) && hasCard(epoch, producer),
    "Snapshot não autoriza estes cartões",
  );
  privateEpoch(g, epoch);
  const snapshot = g.privateState(request.groupId, epoch.id);
  insist(snapshot, "Snapshot privado indisponível");
  verifyGroupSnapshot(snapshot, g.anchor(request.groupId), epoch);
  return { epoch, snapshot };
}

export function snapshotCarrier(
  g: GroupRegistry,
  identity: Identity,
  epoch: GroupEpoch,
  snapshot: GroupSnapshot,
) {
  const material = snapshotMaterial(g, identity.public, epoch, snapshot);
  return sealGroupControl(identity, material.payload, material.readers);
}

export function snapshotMaterial(
  g: GroupRegistry,
  producer: PublicIdentity,
  epoch: GroupEpoch,
  snapshot: GroupSnapshot,
  recipient?: PublicIdentity,
) {
  g.assertScope();
  const anchor = g.anchor(epoch.body.groupId);
  const header = verifyGroupEpoch(epoch, anchor),
    checked = verifyGroupSnapshot(snapshot, anchor, header);
  privateEpoch(g, header);
  insist(
    retainedEpoch(g, anchor.id, header.body.number, header.id).id ===
      header.id && hasCard(header, producer),
    "Só um leitor autorizado pode transportar este snapshot",
  );
  if (recipient)
    insist(hasCard(header, recipient), "Destinatário fora do snapshot");
  return {
    payload: {
      type: "group-control",
      version: 1,
      action: "snapshot",
      groupId: anchor.id,
      epoch: header,
      snapshot: checked,
      ...(recipient ? { to: recipient.id } : {}),
    } as Extract<GroupControl, { action: "snapshot" }>,
    readers: recipient ? [recipient] : checked.members,
  };
}

/** Called only after the epoch header has crossed the runtime's normal
 * authority/stop commit. The carrier signature cannot replace creator proof. */
export function checkedSnapshot(
  g: GroupRegistry,
  bundle: Bundle,
  payload: Extract<GroupControl, { action: "snapshot" }>,
) {
  g.assertScope();
  const { epoch, snapshot } = snapshotPreview(
    bundle,
    payload,
    g.anchor(payload.groupId),
  );
  retainedEpoch(g, payload.groupId, epoch.body.number, epoch.id);
  privateEpoch(g, epoch);
  return { epoch, snapshot };
}

/** Called after openGroupControl. A no-op is proven inside one authenticated
 * scope, never inferred from the carrier ID or a cached UI projection. */
export function alreadyAppliedSnapshot(
  g: GroupRegistry,
  bundle: Bundle,
  payload: Extract<GroupControl, { action: "snapshot" }>,
) {
  const info = g.syncState(payload.groupId);
  if (
    info.view.status !== "active" ||
    !info.admitted ||
    info.checkedThrough === null
  )
    return null;
  let checked: ReturnType<typeof snapshotPreview>;
  try {
    checked = snapshotPreview(bundle, payload, info.anchor);
  } catch {
    return null;
  } // The normal path must still apply any valid safety header.
  if (
    checked.epoch.body.number > info.checkedThrough ||
    g.proofs(payload.groupId, checked.epoch.body.number, 1)[0]?.id !==
      checked.epoch.id
  )
    return null;
  const saved = g.privateState(payload.groupId, checked.epoch.id);
  return saved && canonical(saved) === canonical(checked.snapshot)
    ? checked
    : null;
}

/** Cryptographic preview is not adoption. It can identify a potential proof
 * holder without granting keys or letting an unanchored head win. */
export function snapshotPreview(
  bundle: Bundle,
  payload: Extract<GroupControl, { action: "snapshot" }>,
  anchor: import("./certificates.js").GroupAnchor,
) {
  const epoch = input(() => verifyGroupEpoch(payload.epoch, anchor));
  const snapshot = input(() =>
    verifyGroupSnapshot(payload.snapshot, anchor, epoch),
  );
  insist(
    hasCard(epoch, bundle.manifest.author),
    "Assinante exterior não é leitor do snapshot",
  );
  const expected = payload.to
    ? [...new Set([payload.to, bundle.manifest.author.id])].sort()
    : snapshot.members.map((m) => m.id).sort();
  insist(
    (!payload.to || epoch.body.members.some((m) => m.id === payload.to)) &&
      canonical(bundle.manifest.keys.map((k) => k.reader).sort()) ===
        canonical(expected),
    "Carrier de snapshot alarga leitores",
  );
  return { epoch, snapshot };
}
